### PR TITLE
[1.0.3] Test failure: distributed-transactions-if-test

### LINKED ISSUE #848: Test failure: distributed-transactions-if-test
https://github.com/AntelopeIO/spring/actions/runs/11125907838/job/30915459664

`Duplicate Staring block found: 100 in /__w/spring/spring/build/TestLogs/distributed-transactions-test8/node_bios/stderr.2024_10_01_13_21_16.txt`

Received block 100 late so it restarted speculative block 100. This is expected behavior, so not really sure what we want to do with this. The check was added to guard agaisnt a bug where a block was restarted when it was not suppose to be. Maybe the test should look at the timestamps and only fail if the start block is within some threshold.

### PR DESCRIPTION
**Pull request by greg7mdp**
Resolves #848.

This updates the `verifyStartingBlockMessages` check in `Node.py` to check that we have only one `Starting block` in the log for any block number *unless*:

1. the block was restarted because it was exhausted, 
2. or the second `Starting block` is for a different block time than the first.

I have checked the new regular expression with a small python script against the log and it appears to work fine.

script:
```
import re

def check_file_for_pattern(file_path, pattern):
    regex = re.compile(pattern)

    with open(file_path, 'r') as file:
        for line_num, line in enumerate(file, start=1):
            match = regex.match(line)
            if match:
                blockNumber, time = match.group(1), match.group(2)
                print(f"{blockNumber} -> {time}")

if __name__ == "__main__":
    file_path = "stderr.2024_10_01_13_21_16.txt"
    pattern = r".*Starting block #(\d+) .*(\d\d:\d\d\.\d\d\d) producer"
    check_file_for_pattern(file_path, pattern)
```

execution:

```
❯ python ~/tmp/check_regexp.py
2 -> 21:17.000
3 -> 21:17.500
4 -> 21:18.000
5 -> 21:18.500
6 -> 21:19.000
7 -> 21:19.500
8 -> 21:20.000
9 -> 21:20.500
10 -> 21:21.000
...
```

### CODE REVIEW

**Review by linh2931 — APPROVED** [review]

**Comment by greg7mdp** [comment]
> This should be: "2. or the second `Starting block` is for a different block time than the first."

I don't think so. The full sentence is:

we should have only `Starting block` in the log for any block number **unless** the second `Starting block` is for the exact same block time as the first.

**Comment by heifner** [comment]
> we should have only `Starting block` in the log for any block number **unless** the second `Starting block` is for the exact same block time as the first.

It is actually the opposite of that.

This should be allowed:
```
debug 2024-10-01T13:22:05.393 nodeos    producer_plugin.cpp:2165      start_block          ] Starting block #100 2024-10-01T13:22:06.000 producer defproducerc, deadline 2024-10-01T13:22:05.962
debug 2024-10-01T13:22:06.025 nodeos    producer_plugin.cpp:2165      start_block          ] Starting block #100 2024-10-01T13:22:06.500 producer defproducerc, deadline 2024-10-01T13:22:06.525
```

This should not be allowed:
```
debug 2024-10-01T13:22:05.393 nodeos    producer_plugin.cpp:2165      start_block          ] Starting block #100 2024-10-01T13:22:06.000 producer defproducerc, deadline 2024-10-01T13:22:05.962
debug 2024-10-01T13:22:06.025 nodeos    producer_plugin.cpp:2165      start_block          ] Starting block #100 2024-10-01T13:22:06.000 producer defproducerc, deadline 2024-10-01T13:22:06.525
```

**Review by heifner — APPROVED** [review]

File: tests/TestHarness/Node.py [hunk 1]
```diff
@@ -708,37 +708,36 @@ def countInLog(self, searchStr) -> int:
                 count += contents.count(searchStr)
         return count
 
-    # verify only one or two 'Starting block' per block number unless block is restarted
+    # verify only one 'Starting block' per block number unless block is restarted
```
Comment (linh2931) [inline tests/TestHarness/Node.py:711]: Maybe copy the description in the PR here.

**Review by linh2931 — APPROVED** [review]

**Comment by ericpassmore** [comment]
Note:start
category: Tests
component: Internal
summary: Improve python test harness to better support distributed transactions tests.
Note:end

**Pull request merged**